cpp/src/parquet/schema.cc

```diff
@@ -319,6 +319,35 @@ std::unique_ptr<Node> GroupNode::FromParquet(const void* opaque_element, int nod
                                              params.logical_type, node_id));
 }
 
+namespace {
+
+// If the parquet file is corrupted it is possible the type value decoded
+// will not be in the range of format::Type::type, which is undefined behavior.
+// This method prevents this by loading the value as the underlying type and checking
+// to make sure it is in range.
+inline Type::type SafeLoadType(const format::SchemaElement* element) {
```

Comment:
Is it possible to write this in a less convoluted and more readable way?

Reply:
more readable, yes, at least I've given that a shot by not inline all the template.  Less convoluted ... I'm not sure.  The issue seems to be UBSan is detecting anytime we try to load a enum value that is out of range.  So I have to use the underlying types and memcpy to do the comparisons.  I'm open to suggestions if you have any.

It might have made it less reasonable, but I made the function templated so we could safely load the other enums that were doing this type of conversion.

Reply:
I think it's better now.